test/functional/p2p_private_broadcast.py

```diff
@@ -0,0 +1,390 @@
+            assert_equal(peer.last_message["version"].strSubVer, "/pynode:0.0.1/")
+            assert_equal(peer.last_message["version"].nStartingHeight, 0)
+            assert_equal(peer.last_message["version"].relay, 0)
+            assert_equal(peer.last_message["tx"].tx.rehash(), tx["txid"])
+            self.log.info(f"{label}: ok: outbound connection i={i} is private broadcast of txid={tx['txid']}")
+            broadcasts_done += 1
+
+    def run_test(self):
+        tx_originator = self.nodes[0]
+        tx_receiver = self.nodes[1]
+        far_observer = tx_receiver.add_p2p_connection(P2PInterface())
+
+        wallet = MiniWallet(tx_originator)
+
+        # Fill tx_originator's addrman.
+        for addr in ADDRMAN_ADDRESSES:
+            res = tx_originator.addpeeraddress(address=addr, port=8333, tried=False)
+            if not res["success"]:
+                self.log.debug(f"Could not add {addr} to tx_originator's addrman (collision?)")
+
+        self.wait_until(lambda: len(self.destinations) == NUM_INITIAL_CONNECTIONS)
+
+        # The next opened connection by tx_originator should be "private broadcast"
+        # for sending the transaction. The SOCKS5 proxy should redirect it to tx_receiver.
+
+        txs = wallet.create_self_transfer_chain(chain_length=3)
+        self.log.info(f"Created txid={txs[0]['txid']}: for basic test")
+        self.log.info(f"Created txid={txs[1]['txid']}: for broadcast with dependency in mempool + rebroadcast")
+        self.log.info(f"Created txid={txs[2]['txid']}: for broadcast with dependency not in mempool")
+        tx_originator.sendrawtransaction(hexstring=txs[0]["hex"], maxfeerate=0.1)
+
+        self.log.debug(f"Waiting for outbound connection i={NUM_INITIAL_CONNECTIONS}, "
+                       "must be the first private broadcast connection")
+        self.wait_until(lambda: len(tx_receiver.getrawmempool()) > 0)
+        far_observer.wait_for_tx(txs[0]["txid"])
+        self.log.info(f"Outbound connection i={NUM_INITIAL_CONNECTIONS}: "
+                      "the private broadcast target received and further relayed the transaction")
+
+        # One already checked above, check the other NUM_PRIVATE_BROADCAST_PER_TX - 1 broadcasts.
+        self.check_broadcasts("Basic", txs[0], NUM_PRIVATE_BROADCAST_PER_TX - 1, NUM_INITIAL_CONNECTIONS + 1)
+
+        self.log.info("Resending the same transaction via RPC again")
+        ignoring_msg = f"Ignoring unnecessary request to schedule an already scheduled transaction: txid={txs[0]['txid']}, wtxid={txs[0]['wtxid']}"
+        with tx_originator.busy_wait_for_debug_log(expected_msgs=[ignoring_msg.encode()]):
+            tx_originator.sendrawtransaction(hexstring=txs[0]["hex"], maxfeerate=0.1)
+
+        # How to create a malleated valid witness?
+        #self.log.info("Sending a malleated transaction with a valid witness via RPC")
+        #malleated_valid = malleate_tx(txs[0])
```

Comment:
I don't understand why these lines are commented out. They seem incorrect, since `malleated_valid` is created the exact same way as `malleated_invalid` below. Uncommenting this results in the test failing because `sendrawtransaction` fails with `mandartory-script-verify-flag-failed` but is not caught.

Reply:
I think I understand, the line `How to create a malleated valid witness?` is wondering if there is a way. Then instead of `malleate_tx` we can do the valid way instead.
Maybe that can be made more clear with a comment like: "TODO: Create a malleated valid witness (how?) and substitute `malleate_tx` below?"